pandas/core/groupby/grouper.py

```diff
@@ -947,9 +948,13 @@ def is_in_obj(gpr) -> bool:
             # For the CoW case, we need an equality check as the identity check
             # no longer works (each Series from column access is a new object)
             try:
-                return gpr.equals(obj[gpr.name])
+                obj_gpr_column = obj[gpr.name]
             except (AttributeError, KeyError, IndexError, InvalidIndexError):
                 return False
+            if isinstance(gpr, Series) and isinstance(obj_gpr_column, Series):
+                ref = weakref.ref(obj_gpr_column._mgr.blocks[0])
+                return ref in gpr._mgr.blocks[0].refs.referenced_blocks
```

Comment:
So if you create multiple times a weakref to the same object, this is always the same (i.e. identical) weakref object?